### DISCUSSION TITLE
Cannot login from iPhone

### DISCUSSION THREAD

**Discussion by sbogomolov** [9p4/jellyfin-plugin-sso#162] [Q&A] [answered]
I'm using authentik and can login from my laptop just fine. When I try to login from iPhone - I get redirected to authentik, authentication happens successfully, and I get redirected back to jellyfin start page (not logged in). I tried with Safari and Chrome. The log looks fine, no errors:
```
[2024-02-05 01:15:56.700 +01:00] [INF] [64] Jellyfin.Plugin.SSO_Auth.Api.SSOController: SSO Controller initialized
[2024-02-05 01:15:57.787 +01:00] [WRN] [70] Jellyfin.Server.Middleware.ResponseTimeMiddleware: Slow HTTP Response from "https://jellyfin.example.com/sso/OID/start/authentik" to "<ipv4_was_here>" in 0:00:01.0877651 with Status Code 302
[2024-02-05 01:15:58.567 +01:00] [INF] [70] Jellyfin.Plugin.SSO_Auth.Api.SSOController: SSO Controller initialized
[2024-02-05 01:16:01.006 +01:00] [INF] [70] Jellyfin.Plugin.SSO_Auth.Api.SSOController: Is request linking: False
[2024-02-05 01:16:01.007 +01:00] [WRN] [70] Jellyfin.Server.Middleware.ResponseTimeMiddleware: Slow HTTP Response from "https://jellyfin.example.com/sso/OID/redirect/authentik?code=<code_was_here>" to "<ipv4_was_here>" in 0:00:02.4403616 with Status Code 200
[2024-02-05 01:16:01.124 +01:00] [INF] [70] Jellyfin.Plugin.SSO_Auth.Api.SSOController: SSO Controller initialized
[2024-02-05 01:16:01.147 +01:00] [INF] [70] Jellyfin.Plugin.SSO_Auth.Api.SSOController: Auth request created...
[2024-02-05 01:16:01.148 +01:00] [INF] [70] Emby.Server.Implementations.Session.SessionManager: Current/Max sessions for user "<username>": 4/0
[2024-02-05 01:16:01.153 +01:00] [INF] [70] Emby.Server.Implementations.Session.SessionManager: Reissuing access token: "<token_was_here>"
```

I am able to login from my laptop. I am using Authentik for other services and these work fine from all devices. What could be wrong here?

**Comment by 9p4** [comment]
You don't seem to be the first person to have an issue with the iOS/Authentik/Jellyfin combination: https://github.com/9p4/jellyfin-plugin-sso/issues/113

**Reply by sbogomolov** [reply]
Thanks for the link. I have seen that issue. It looks like eventually the focus there shifted towards `endpoint belongs to different authority` problem, which I have already resolved. Is there anything I can do to debug this issue? Should I post in the linked issue?

**Reply by 9p4** [reply]
Do you have a mac with which you can debug the Safari iOS responses?

**Reply by sbogomolov** [reply]
Yes, I do have a Mac. Could you leave some pointers as to what needs to be done and I'll try to figure it out.

**Reply by 9p4** [reply]
Take a look at https://developer.apple.com/documentation/safari-developer-tools/inspecting-ios

**Reply by sbogomolov** [reply]
I think I have found an issue. At first glance, everything looks fine:
- I see request to `https://jellyfin.example.com/sso/OID/start/authentik`
- It gets redirected to `https://auth.example.com/application/o/authorize`
- We get back 200 and redirect to `https://jellyfin.example.com/sso/OID/redirect/authentik`
- Then we are redirected to `https://jellyfin.example.com/web/index.html`

I have noticed that `https://jellyfin.example.com/web/index.html` is loaded from cache. If I **Disable Caches** - authentication works.

Another observation: when Jellyfin is restarted - authentication works once. If I logout and try to login again - it does not work.

I have tried to add query parameters (timestamps) [here](https://github.com/9p4/jellyfin-plugin-sso/blob/db203808f29bfe75047d8d295bd033d0166e1a14/SSO-Auth/WebResponse.cs#L496) and [here](https://github.com/9p4/jellyfin-plugin-sso/blob/db203808f29bfe75047d8d295bd033d0166e1a14/SSO-Auth/WebResponse.cs#L504). This resulted in `https://jellyfin.example.com/web/index.html` not being loaded from cache. This did not fix the problem though.

It is definitely related to caches because **Disable Caches** does the trick. Could it be that some other resource (like .js file) is dynamically generated and caching that results in a failure?

**Reply by 9p4** [reply]
It might be the redirect response that is being cached. You have no idea how much this helps me :)

**Reply by 9p4** [reply]
Would you be willing to compile and install a custom version of the plugin to test whether a fix would work?

**Reply by sbogomolov** [reply]
Sure, just point me to the branch.

**Reply by 9p4** [reply]
Could you try the `cache-buster` branch?

**Reply by sbogomolov** [reply]
I have tried it. I see the `Cache-Control` cookie on the response, but this did not work. I have experimented a bit more and found something. There are 4 keys in the local storage:
- `_deviceId2`
- `enableAutoLogin`
- `jellyfin_credentials`
- `user-<hash>`

When I sign-out - the `user-<hash>` key is removed from the local storage and following attempts to authenticate do not work. However, if I remove `jellyfin_credentials` or `_deviceId2` key after signing out - I can sign back in. It looks like absence of any of these two keys forces new `jellyfin_credentials` to be created. ~I guess removal of these keys when signing out is not in this plugin's scope. But maybe you can force re-creation of `jellyfish_credentials`?~ I see that you are updating it every time already.

**Reply by sbogomolov** [reply]
Another update: this problem is also present in Safari on Mac. I have checked the value of `jellyfin_credentials` when trying to login. After first login - it looks OK. After logout - it does not change the value. When trying to login the second time (and that fails) - `AccessToken` and `UserId` in `jellyfin_credentials` are set to `null`.

**Reply by sbogomolov** [reply]
Did a bit more experiments. Added `no-cahe, no-store` to `/SSO/Auth/...` request. This did not fix it. Then I added logging to your JS code that sets `jellyfin_credentials`. I have printed the value before change and after change. Then I checked the console and saw following:
- Value before update: contains OLD_TOKEN
- Value after update: contains NEW_TOKEN

Looks good, however right after that I saw another log from `main.jellyfin.bundle.js:34284`. That message was `Stored JSON credentials` and contained OLD_TOKEN. It looks like somehow, somewhere it was cached and overwritten after your JS code updated the local storage.

**Reply by 9p4** [reply]
I think I need to add something to the JS to erase the values set in the local storage, then load the iframe, then wait for the localstorage values to get repopulated, then continue with authentication.

**Reply by sbogomolov** [reply]
This is exactly what I was thinking as well :) I'll try and let you know how it went.

**Reply by sbogomolov** [reply, marked as answer]
This fixes it: https://github.com/9p4/jellyfin-plugin-sso/pull/163.
Tested with Safari on Mac and iPhone.
